### PR TITLE
Exercise1

### PR DESCRIPTION
**Pull request by Tobisaninfo**

### CODE REVIEW

File: exercises/1/src/components/HackerNews.vue [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        voteCount: 0,
+      }],
+      id: 1
+    }
+  },
+  computed: {
+    sortedItems() {
+      let copiedItems = [...this.items];
+      copiedItems.sort((o1, o2) => o2.voteCount - o1.voteCount)
+      return copiedItems;
```
Comment (medizinmensch) [inline exercises/1/src/components/HackerNews.vue]: ⭐ for sorted properties - even as computed property! Nice!

In the future, you could shorten like this:

```suggestion
      return [...this.items].sort((o1, o2) => o2.voteCount - o1.voteCount);
```

File: exercises/1/src/components/HackerNews.vue [hunk 2]
```diff
@@ -0,0 +1,69 @@
+<template>
+  <div class="hacker-news">
+    <h1>News List</h1>
+    <div v-for="item in sortedItems" :key="item.id">
+      <NewsItem v-bind:news-item="item" v-on:news_remove="onNewsRemove"/>
+    </div>
```
Comment (medizinmensch) [inline exercises/1/src/components/HackerNews.vue]: ⭐ for title and votes list

File: exercises/1/src/components/HackerNews.vue [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      let newsItem = {
+        id: this.id++,
+        title: newsTitle,
+        voteCount: 0
+      };
+      this.items.push(newsItem);
+    },
+    onNewsRemove(newsItem) {
+      this.items = this.items.filter((element) => element.id !== newsItem.id)
+    },
```
Comment (medizinmensch) [inline exercises/1/src/components/HackerNews.vue]: ⭐ for "remove functionality" and emitting parent component which then removes the item.

File: exercises/1/src/components/NewsItem.vue [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  data() {
+    return {}
+  },
+  methods: {
+    upvote() {
+      this.newsItem.voteCount += 1;
+    },
+    downvote() {
+      this.newsItem.voteCount -= 1;
+    },
```
Comment (medizinmensch) [inline exercises/1/src/components/NewsItem.vue]: As described in the exercise description, child components should use `$emit`to update the item in the parent component. This was not done here, so I cant give you the point on this one.

❌

File: exercises/1/src/components/NewsForm.vue [hunk 5]
```diff
@@ -0,0 +1,29 @@
+<template>
+  <div id="news-input">
+    <label>
+      <input v-model="news_input" placeholder="edit me">
+    </label>
+    <button @click="createNewsItem">create</button>
```
Comment (medizinmensch) [inline exercises/1/src/components/NewsForm.vue]: Using a button like this works. But the task was to use a `<form>` and preventing it's default behavior (POSTing). This, I can not give you the point for this one. 

❌

**Review by medizinmensch — APPROVED** [review]
You got 3 stars in the review below. Furthermore:

* you pulled the new exercises :star: 
* using functional programming paradigms :star: 
* there are no runtime errors :star:

Sadly, we found many unrelated changes. For instance:
```
exercises/1/.editorconfig
exercises/1/build/* (around 8 files)
```

Thus, we can't give you a point on this one ❌. Fell free to remove those files. On the other hand, having the package-lock.json would be very nice. This way, you can make sure a dev that uses your build get's the exact same npm package versions. 

One point was for "Components have clean responsibilities". I can not give you the point for that.
The paradigma goes "Data down, Events up".
I see that you passed down news-item data into your props. But props should not be changed. And this data should be managed in a central point anyway. Your `HackerNews.vue` would be suitable for that. For that, you would have to emit events to that component and let that component change the data. 
Thus, I can not give you the ⭐ for `⭐ for "Components have clean responsibilities" ❌ 


You got 6/10 in total ⭐

File: exercises/1/src/components/NewsItem.vue (hunk 4 above)
Comment (Tobisaninfo) [reply exercises/1/src/components/NewsItem.vue]: I see your point and we had implemented it your suggested way. But last week, Robert told us in the practice time, that it is ok, handling the vote count directly in the component itself. So thats the reason, why we shifted this code into that section

File: exercises/1/src/components/NewsForm.vue (hunk 5 above)
Comment (Tobisaninfo) [reply exercises/1/src/components/NewsForm.vue]: We traded this Sentence `Prevent the default behaviour of a <form> html element which would execute a HTTP request on submit.` like "If you use a form, take care of the redirect aspect" and not like you say "You have to use a form tag regardless what ever options are possible". So our point of view is, this sentence is more like a hint such as the sort side-effect thing.

**Comment by Tobisaninfo** [comment]
We think all files like ./build or other files are needed, because either the project is not runnable without these files or because they are part of this hole project. The exercise included for us the complete setup of a fresh vue project, that is runnable without much effort. This task was not like contributing to an existing project, where these files might irrelevant or even exists already.

File: exercises/1/src/components/NewsItem.vue (hunk 4 above)
Comment (roschaefer) [reply exercises/1/src/components/NewsItem.vue]: @medizinmensch and @Tobisaninfo are both correct. I made a mistake with my own reference implementation and updated the props in the child component. There is an easy way how to solve that:
a) You create a new state based on a prop in `data()`
b) You simply create a new object on-the-fly:
```js
this.$emit('update', { ...this.newsItem, voteCount: (this.newsItem.voteCount + 1)})
```

File: exercises/1/src/components/NewsForm.vue (hunk 5 above)
Comment (roschaefer) [reply exercises/1/src/components/NewsForm.vue]: Then this is a weakness of the exercise description:
```
:star: Below the list is a form to add new items to the list
```
should be
```
:star: Below the list is a <form> to add new items to the list
```
indicating that a `<form>` html is required here.

File: exercises/1/build/build.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
+'use strict'
```
Comment (roschaefer) [inline exercises/1/build/build.js]: The moral-message of this task:
```
:star: Your PR does not contain unrelated changes
- Double check "Files changed" tab in your pull request
```
was to get empathy for your code reviewer :laughing:

You're asking your co-worker to look at your changed files. *This takes time.* As the person who is requesting a code review, you should learn to put only relevant changes in the PR.

In your job, you will receive the best code reviews when your "Files changed" are *small*.

File: exercises/1/build/build.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      process.exit(1)
+    }
+
+    console.log(chalk.cyan('  Build complete.\n'))
+    console.log(chalk.yellow(
+      '  Tip: built files are meant to be served over an HTTP server.\n' +
+      '  Opening index.html over file:// won\'t work.\n'
+    ))
+  })
+})
```
Comment (roschaefer) [inline exercises/1/build/build.js]: I'm also missing documentation how to run your code. Please put instructions in your `README.md`. That's why I suggested to use a mono-repo directory layout in the exercise description and you put your code into a folder `webapp/` along with a `README.md`.

**Review by roschaefer — CHANGES_REQUESTED** [review]
progress

Agree with @medizinmensch here!

File: exercises/1/src/components/NewsItem.vue (hunk 4 above)
Comment (roschaefer) [reply exercises/1/src/components/NewsItem.vue]: for solution a) see here: https://vuejs.org/v2/guide/components-props.html#One-Way-Data-Flow

Comment (Tobisaninfo) [reply exercises/1/src/components/NewsItem.vue]: Thanks for your advice. We decided to implement an additional `$emit` method.

**Pull request merged**